### PR TITLE
deltatouch: init at 1.4.0

### PR DESCRIPTION
**Pull request by link2xt**
## Description of changes

This is an initial commit for [DeltaTouch](https://codeberg.org/lk108/deltatouch) Delta Chat client.
It is a [Delta Chat](https://delta.chat) client for Ubuntu Touch, [available at OpenStore](https://open-store.io/app/deltatouch.lotharketterer).  

## Things done

Built on x86_64-linux (Arch Linux with Nix installed via [Determinate Nix Installer](https://github.com/DeterminateSystems/nix-installer)).

Having correct icons requires installing qt5ct in the profile and selecting "suru" icon theme.
Tested by running:
```
$ nix profile install nixpkgs#qt5ct nixpkgs#lomiri.suru-icon-theme
$ export QT_QPA_PLATFORMTHEME=qt5ct
$ qt5ct # select Suru icon theme
$ nix run github:guibou/nixGL#nixGLIntel -- nix shell .#deltatouch
```

This is how it looks like:

It is possible to configure account, send and receive messages. File selection dialog currently does not work outside of Lomiri desktop environment and I am not running the whole Lomiri, but it is usable anyway.

### CODE REVIEW

**Comment by OPNA2608** [comment]
Lomiri in NixOS is 1-2 PRs away so I wanna give this a look, will give a review when I have some spare time.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:51]: For the wrapper to automatically add QML import, they need to be installed to `$out/${qtbase.qtQmlPrefix}` (which is `lib/qt-${qtbase.version}/qml`). Would it be possible to patch the install location in the CMake script instead?

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:64]: I think icons are one of the few things we want to be impurely-supplied to the environment, so different icon themes can be used without rebuilding everything.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:75]:
```suggestion
    maintainers = with maintainers; [ link2xt ];
```

And please preceed the init with a separate commit that adds you to [`/maintainers/maintainers-list.nix`](https://github.com/NixOS/nixpkgs/blob/9d64752753ec04998437b1a3ca57d27b66c47891/maintainers/maintainer-list.nix). Read the comment at the top of that file for notes on how to do it.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:77]:
```suggestion
    platforms = platforms.linux;
```

Unless there's also supposed to be support for Windows, macOS, embedded systemd etc. Though most of the Lomiri stuff doesn't support those platforms either. :wink:

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Welcome to Nixpkgs! Haven't tried to build yet, but here are some initial notes.

Also to consider:

- `finalAttrs` style instead of `rec` style. Just grep Nixpkgs for some examples.
- Can we do without needing to copy-paste libraries into the build & output tree? Either patching the CMake script to link against system-supplied versions, or at least replacing the ones in the output tree with symlinks to the originals?

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [reply pkgs/by-name/de/deltatouch/package.nix]: That's all QML modules, so should stay.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:63]: It's a (soft?) convention to have `buildInputs` right after `nativeBuildInputs`. Could you move this up to `nativeBuildInputs`?

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Launches fine:
Bildschirmfoto_2024-03-01_14-54-32

With the suggested changes, localisation works partially (large parts seem untranslated):
Bildschirmfoto_2024-03-01_15-51-02

File: pkgs/by-name/de/deltatouch/package.nix
Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:51]: You're correct on all accounts. It was needed at some point in the past, and it may soon be needed again as I will try to merge the nixpkg branch into main again.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:69]: This could prolly still be done better… Either submitted upstream somehow (and tested in the click usecase) as mentioned earlier, or done as a vendored patch (which would require occasional fixing on bumps, and maybe some shuffling around to get the correct `$out` in there)…

Prolly fine for now, but I'll put giving this another look onto my Lomiri TODO list if you can't take a look at tackling this.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:38]: Readme of the Nixpkgs branch / that commit says:

> […] Differences to the Ubuntu Touch version are:
> - […]
> - The Lomiri Connectivity module is not used

Is this needed here, is the readme outdated?

Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:38]: I'm sorry, the readme is outdated, will fix asap

Comment (OPNA2608) [reply pkgs/by-name/de/deltatouch/package.nix:38]: Should we switch this to the main branch then?

Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:38]: Yes. I had not informed link2xt about this because I did not have time before to test whether it runs from main. I just checked, it does.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:69]: I can try to add it to Main.qml, but I'm not sure how to handle the "placeholder out" part.

Comment (OPNA2608) [reply pkgs/by-name/de/deltatouch/package.nix:69]: Some Lomiri stuff does it by via a context property, which sets it via a define in C/C++ land:

https://gitlab.com/ubports/development/core/lomiri-system-settings/-/blob/e21e51a333f8f285f97953abc03b970df149fadf/src/qml/MainWindow.qml#L77
```qml
    Component.onCompleted: {
        i18n.domain = "lomiri-system-settings"
        i18n.bindtextdomain("lomiri-system-settings", i18nDirectory)
```
https://gitlab.com/ubports/development/core/lomiri-system-settings/-/blob/e21e51a333f8f285f97953abc03b970df149fadf/src/main.cpp#L86
```c++
    view.rootContext()->setContextProperty("i18nDirectory", mountPoint + I18N_DIRECTORY);
```

https://gitlab.com/ubports/development/core/lomiri-system-settings/-/blob/e21e51a333f8f285f97953abc03b970df149fadf/src/CMakeLists.txt#L3
```cmake
add_definitions(-DI18N_DIRECTORY="${CMAKE_INSTALL_PREFIX}/share/locale")
```
(This should be `CMAKE_INSTALL_FULL_LOCALEDIR`, fix has been submitted to them.)

---

Alternatively, you could turn the file into a template:
```qml
    Component.onCompleted: {
        i18n.domain = "deltatouch.lotharketterer"
        i18n.bindtextdomain("deltatouch.lotharketterer", "@CMAKE_INSTALL_FULL_LOCALEDIR@")
```
and have CMake replace the placeholder at build time:
```cmake
configure_file(${CMAKE_CURRENT_SOURCE_DIR}/Main.qml.in ${CMAKE_CURRENT_BINARY_DIR}/Main.qml @ONLY)
```

---

Either method would also require using
```cmake
include(GNUInstallDirs)
```
to have proper guarantees on `CMAKE_INSTALL_FULL_LOCALEDIR`, which also opens up further fix upstreaming for install locations from the `postPatch` script.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix]: Turns out that even after merging it into main, it's not needed anymore to set CLICKABLE_DESKTOP_MODE, so this part can be removed

File: pkgs/by-name/de/deltatouch/package.nix
Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:69]: Thanks, I'll have a look

File: pkgs/by-name/de/deltatouch/package.nix
Comment (dotlambda) [inline pkgs/by-name/de/deltatouch/package.nix:78]:
```suggestion
    changelog = "https://codeberg.org/lk108/deltatouch/src/commit/${src.rev}/CHANGELOG";
    description = "Messaging app for Ubuntu Touch, powered by Delta Chat core";
```
I know it's the official description but I'm not sure mentioning Ubuntu Touch makes sense in Nixpkgs.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (dotlambda) [inline pkgs/by-name/de/deltatouch/package.nix:59]: Does
```suggestion
      --replace-fail 'IMPORTED_LOCATION "''${CMAKE_CURRENT_BINARY_DIR}/libquirc.so.1.2"' 'IMPORTED_LOCATION "${lib.getLib quirc}/lib/libquirc.so"'
```
work?

**Comment by dotlambda** [comment]
I think the version should be `1.3.1-unstable-...` because 1.3.1 is the last version from before the commit.

**Comment by lk108** [comment]
> > I think the version should be `1.3.1-unstable-...` because 1.3.1 is the last version from before the commit.
> 
> There is a "prepare 1.4.0" commit. The tag is missing. cc @lk108 did you forgot to push the tag or will you tag some later commit with 1.4.0?

I'm sorry, the tag should be there now.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (dotlambda) [reply pkgs/by-name/de/deltatouch/package.nix:59]: > This is something to fix in libquirc package I guess?

Indeed. That package also ships a static library by default which it should not.
cc @7c6f434c @4825764518 @rnhmjoj

Comment (7c6f434c) [reply pkgs/by-name/de/deltatouch/package.nix:59]: I mostly care about the executables so I follow upstream defaults for `make install` w.r.t. libraries, but if this PR adds a `libquirc.so` symlink to `quirc`, it's perfectly fine.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:69]: With the first option you suggested, it works on Ubuntu Touch, but not on Nix. [Here's](https://codeberg.org/lk108/deltatouch/commit/b650643ae90a72a3d699eaf374094d279e8d7972) the corresponding commit.

On Nix, the define in C++ as set by cmake is `/nix/store/2az7rny9qv110fk57paxqwvpy7hj8l3l-deltatouch-1.4.0-unstable-2024-03-08/share/locale`, which actually contains the lang dirs like `de/LC_MESSAGES/deltatouch.lotharketterer.mo`. Yet the translations are not used.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (dotlambda) [reply pkgs/by-name/de/deltatouch/package.nix:59]: fixed in https://github.com/NixOS/nixpkgs/pull/294589

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [reply pkgs/by-name/de/deltatouch/package.nix:69]: > it works on Ubuntu Touch, but not on Nix

Seems to work fine for me. Not sure what state the application is stuck in on my end, so I had to add a translation for it to show up (immediately jumps to this "Known accounts" screen that I can't exit), but the catalogue gets loaded & can be used with that commit.

image

Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:69]: It's merged into main, commit dcfdd8a0fca5fff10d0383f77f4c0cbea302de00. In package.nix, the `substituteInPlace po/CMakeLists.txt`and `substituteInPlace qml/Main.qml` parts should now be removed, right?

For some reason, for me on Nix, `QLocale::system().name()` in C++ returns de_DE, but `i18n.language` in Main.qml is empty, so it doesn't even try to open any .mo file.

Comment (OPNA2608) [reply pkgs/by-name/de/deltatouch/package.nix:69]: > You don't have the icons set up

Gah, right. Laziness be damned… :sweat_smile: 

> In package.nix, the `substituteInPlace po/CMakeLists.txt` and `substituteInPlace qml/Main.qml` parts should now be removed, right?

Yes.

> `QLocale::system().name()` in C++ returns de_DE, but `i18n.language` in Main.qml is empty.

Hmm. You might need to look into how LUITK sets this, and what environment variables are used to determine it: https://gitlab.com/ubports/development/core/lomiri-ui-toolkit/-/blob/355b2c09a64efddc80b7ea9fbcc9d1462d7573cf/src/LomiriToolkit/i18n.cpp#L62-71

Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:69]: > Hmm. You might need to look into how LUITK sets this, and what environment variables are used to determine it

Thanks for the link. According to the comments in i18n.cpp, it should work with my env. For the moment, I am out of ideas. Do you consider it a blocker if this is not resolved?

Comment (OPNA2608) [reply pkgs/by-name/de/deltatouch/package.nix:69]: Bummer. :slightly_frowning_face: 

I think we should at least rewrite this substitution here as a vendored patch then, due to its insertion of multiple lines.

Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix:69]: Well.. now translations don't show up with the substitution either. Something must be wrong with my setup.

**Comment by dotlambda** [comment]
How do I add a new account?
image

**Comment by lk108** [comment]
> How do I add a new account?

You don't see the necessary icons, quote from link2xt above:
$ nix profile install nixpkgs#qt5ct nixpkgs#lomiri.suru-icon-theme
$ export QT_QPA_PLATFORMTHEME=qt5ct
$ qt5ct # select Suru icon theme

**Comment by lk108** [comment]
Actually, the "+" icon is used to set up a new account

**Review by dotlambda — APPROVED** [review]
Works well with the hints. I didn't test translations etc though.

Selecting the Suru icon theme in qt5ct was not necessary btw.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [reply pkgs/by-name/de/deltatouch/package.nix:69]: FWIW, applying https://codeberg.org/lk108/deltatouch/commit/dcfdd8a0fca5fff10d0383f77f4c0cbea302de00 worksfine for me. Might be a setup issue then?

image

I'll send a suggestion to fetch that as a patch & drop/adjust the substitutions affected by it, then this LGTM.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:3]:
```suggestion
, fetchFromGitea
, fetchpatch
```

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:25]:
```suggestion

  patches = [
    (fetchpatch {
      name = "0001-deltatouch-Fix-localisation.patch";
      url = "https://codeberg.org/lk108/deltatouch/commit/dcfdd8a0fca5fff10d0383f77f4c0cbea302de00.patch";
      hash = "sha256-RRjHG/xKtj757ZP2SY0GtWwh66kkTWoICV1vDkFAw3k=";
    })
  ];

  nativeBuildInputs = [
```

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:48]: Above patch adds GNUInstallDirs, so we can use them here now.

```suggestion
      --replace-fail 'set(DATA_DIR /)' 'set(DATA_DIR ''${CMAKE_INSTALL_DATAROOTDIR})' \
      --replace-fail 'RUNTIME DESTINATION ''${CMAKE_INSTALL_PREFIX}' 'RUNTIME DESTINATION ''${CMAKE_INSTALL_BINDIR}' \
      --replace-fail 'assets/logo.svg DESTINATION assets' 'assets/logo.svg DESTINATION ''${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/scalable/apps RENAME deltatouch.svg' \
      --replace-fail "\''${DESKTOP_FILE_NAME} DESTINATION \''${DATA_DIR}" "\''${DESKTOP_FILE_NAME} DESTINATION \''${CMAKE_INSTALL_DATAROOTDIR}/applications"
```

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:55]: Included in the patch.

```suggestion
```

File: pkgs/by-name/de/deltatouch/package.nix
Comment (OPNA2608) [inline pkgs/by-name/de/deltatouch/package.nix:69]: Included in the patch.

```suggestion
```

**Review by OPNA2608 — COMMENTED** [review]
If this is fine with you, considering there was some uncertainty about the patch.

File: pkgs/by-name/de/deltatouch/package.nix
Comment (lk108) [reply pkgs/by-name/de/deltatouch/package.nix]: > I'll send a suggestion to fetch that as a patch & drop/adjust the substitutions affected by it, then this LGTM. That fine with you?

Not sure if you're asking me, but I'm fine with it :-)

**Review by OPNA2608 — APPROVED** [review]
Result of `nixpkgs-review pr 289642` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>deltatouch</li>
  </ul>
</details>

---

LGTM! Thanks for sticking through with all the reviews and back n forth!

**Pull request merged**
